unityx3d.cs

```diff
@@ -185,259 +226,435 @@ public static float ToDegrees(float radians)
+        class TransformInfo
         {
-            return Tools.Vector3FromString(GetAttribute(node, attrib), defaultScalar);
+            public Vector3 translation, scale;
+            public Quaternion rotation;
         }
 
-        static Vector4 Vector4FromAttribute(XmlNode node, string attrib, float defaultScalar = 0)
-        {
-            return Tools.Vector4FromString(GetAttribute(node, attrib), defaultScalar);
-        }
+        //
+        // Map to help implement DEF/USE in X3D's "Shape" nodes
+        //
 
-        static float FloatFromAttribute(XmlNode node, string attrib)
-        {
-            string v = GetAttribute(node, attrib);
+        Dictionary<string, GameObject> DefToObjMap = new Dictionary<string, GameObject>();
 
-            if (v == "")
-                return 0;
+        //
+        // Extract node name (if present), populating DefToObj map as we go.
+        //
 
-            return float.Parse(v);
+        string GetName(XmlNode node, GameObject obj)
+        {
+            if (node.Attributes != null)
+            {
+                XmlAttribute name = node.Attributes["DEF"];
+                if (name != null)
+                {
+                    DefToObjMap[name.Value] = obj;
+                    return name.Value;
+                }
+            }
+
+            return node.Name ?? "Unnamed";
         }
 
-        static void ReadBox(XmlNode boxNode, GameObject obj)
+        //
+        // Mesh generation 
+        //
+
+        GameObject ReadBox(XmlNode boxNode, GameObject parent)
         {
-            Vector3 size = Vector3FromAttribute(boxNode, "size", 1);
+            Vector3 size = Tools.Vector3FromAttribute(boxNode, "size", 1);
+            GameObject cube = GameObject.CreatePrimitive(PrimitiveType.Cube);
+
+            GameObject obj = new GameObject("Box");
+
+            obj.transform.SetParent(parent.transform);
             obj.transform.localScale = size;
 
-            GameObject cube = GameObject.CreatePrimitive(PrimitiveType.Cube);
             MeshFilter meshFilter = obj.AddComponent<MeshFilter>();
             meshFilter.mesh = cube.GetComponent<MeshFilter>().sharedMesh;
+
             GameObject.DestroyImmediate(cube);
+
+            return obj;
         }
 
-        static void ReadIndexedFaceSet(XmlNode indexedFaceSetNode, GameObject obj)
+        List<GameObject> ReadIndexedFaceSet(XmlNode indexedFaceSetNode, GameObject parent)
         {
-            int[] coordIndex = Tools.IntArrayFromString(GetAttribute(indexedFaceSetNode, "coordIndex"));
-            if (coordIndex.Length == 0)
-                return;
-
-            XmlNode coordinateNode = null;
-            XmlNode normalNode = null;
-
-            foreach(XmlNode child in indexedFaceSetNode)
+            List<GameObject> components = new List<GameObject>();
+            float[] _vtx = null, _uv = null, _nrm = null, _rgba = null;
```

Comment:
In favor of a single style for naming things, can you change these to CamelCase names?